File: firmware/lib/split42/split42.cpp
#include "split42.h"

KBHalf::KBHalf(uint8_t addr, std::initializer_list<uint8_t> pinmap,
               std::initializer_list<uint8_t> keymap)
    : addr(addr) {
    // pinmap
    assert(pinmap.size() <= 21);
    int i = 0;
    for (auto pin : pinmap) {
        assert(pin < 24);
        this->pinmap[pin] = i++;
    }

    // keymaps
    assert(keymap.size() <= 21);
    i = 0;
    for (auto key : keymap) {
        this->keymap[i++] = key;
    }
}

void KBHalf::init() {
    write(0x10, 0xFF, 5); // invert all polarity
    write(0x18, 0xFF, 5); // set all to inputs
    write(0x20, 0x00, 5); // enable interrupts on all pins
    get_state();          // clear all interupts
}

void KBHalf::write(uint8_t reg, uint8_t value, uint8_t len) {
    Wire.beginTransmission(addr);
    Wire.write(reg | 0x80); // add auto-incrementing
    for (uint8_t i = 0; i < len; ++i)
        Wire.write(value);
    Wire.endTransmission();
}

uint32_t KBHalf::get_state() {
    Wire.beginTransmission(addr);
    Wire.write(0x80);
    Wire.endTransmission(false);
    uint8_t buf[5];
    uint8_t count = Wire.requestFrom(addr, (uint8_t)5);
    for (uint8_t i = 0; i < count; ++i)
        buf[i] = Wire.read();

    return (buf[0]) | (buf[3] << 8) | (buf[4] << 16);
}

File: firmware/split42.py
from kmk.scanners import Scanner
from keypad import Event as KeyEvent

class split42(Scanner):
    def __init__(self, i2c, interrupt, l_addr, r_addr, banks, pin_map):
        super().__init__()
        self.i2c = i2c
        self.interrupt = interrupt
        self.l_addr = l_addr
        self.r_addr = r_addr
        self.banks = banks
        self.pin_map = pin_map
        self._key_count = len(pin_map)
        
        self.interval = 0.01

        self.prev = 0
        self.buffer = []
        
        while not i2c.try_lock():
            pass
        try:
            # set polarity, set inputs, enable interrupts
            for reg, val in [(0x10, 0xFF), (0x18, 0xFF), (0x20, 0x00)]:
                data = bytes([reg | 0x80] + [val] * 5)
                    
                i2c.writeto(l_addr, data)
                i2c.writeto(r_addr, data)
                
            # read to clear interrupt
            buf = bytearray(5)
            self.i2c.writeto_then_readfrom(self.l_addr, bytes([0x80]), buf)
            self.i2c.writeto_then_readfrom(self.r_addr, bytes([0x80]), buf)
        finally:
            i2c.unlock()
    # end __init__
        
    @property
    def key_count(self):
        return self._key_count
    # end key_count

    def scan_for_changes(self):
        if (self.interrupt.value == 1) and (not self.buffer): return None
        
        while not self.i2c.try_lock():
            pass
        try:
            l_data = bytearray(5)
            r_data = bytearray(5)
            self.i2c.writeto_then_readfrom(self.l_addr, bytes([0x80]), l_data)
            self.i2c.writeto_then_readfrom(self.r_addr, bytes([0x80]), r_data)
        finally:
            self.i2c.unlock()

        l_data = [l_data[bank] for bank in self.banks]
        r_data = [r_data[bank] for bank in self.banks]

        data = bytes(l_data + r_data)
        state = int.from_bytes(data, 'big')
        
        pressed = state & ~self.prev
        released = ~state & self.prev
        self.prev = state

        for idx, pin in enumerate(self.pin_map):
            if bool(pressed & (1 << pin)):
                self.buffer.append(KeyEvent(idx, True))
            elif bool(released & (1 << pin)):
                self.buffer.append(KeyEvent(idx, False))

        if self.buffer:        
            return self.buffer.pop(0)
        return None
    # end scan_for_changes